File: context_tree_cli/rules/root_node.py
"""Rule: check root NODE.md exists and has valid frontmatter."""

from __future__ import annotations

from context_tree_cli.repo import Repo


def evaluate(repo: Repo) -> dict:
    tasks = []
    if not repo.path_exists("NODE.md"):
        tasks.append(
            "NODE.md is missing — create from"
            " `.context-tree/templates/root-node.md.template`,"
            " fill in your project's domains"
        )
    else:
        fm = repo.frontmatter("NODE.md")
        if fm is None:
            tasks.append(
                "NODE.md exists but has no frontmatter —"
                " add frontmatter with title and owners fields"
            )
        else:
            if not fm.get("title") or fm["title"].startswith("<"):
                tasks.append(
                    "NODE.md has a placeholder title — replace with your"
                    " organization name"
                )
            if not fm.get("owners") or (
                len(fm["owners"]) == 1 and fm["owners"][0].startswith("<")
            ):
                tasks.append(
                    "NODE.md has placeholder owners — set owners to your"
                    " GitHub username(s)"
                )
        if repo.has_placeholder_node():
            tasks.append(
                "NODE.md has placeholder content — fill in your project's"
                " domains and description"
            )
    return {"group": "Root Node", "order": 2, "tasks": tasks}

File: src/rules/root-node.ts
import type { Repo } from "../repo.js";
import type { RuleResult } from "./index.js";

export function evaluate(repo: Repo): RuleResult {
  const tasks: string[] = [];
  if (!repo.pathExists("NODE.md")) {
    tasks.push(
      "NODE.md is missing — create from `.context-tree/templates/root-node.md.template`, fill in your project's domains",
    );
  } else {
    const fm = repo.frontmatter("NODE.md");
    if (fm === null) {
      tasks.push(
        "NODE.md exists but has no frontmatter — add frontmatter with title and owners fields",
      );
    } else {
      if (!fm.title || fm.title.startsWith("<")) {
        tasks.push(
          "NODE.md has a placeholder title — replace with your organization name",
        );
      }
      if (
        !fm.owners ||
        fm.owners.length === 0 ||
        (fm.owners.length === 1 && fm.owners[0].startsWith("<"))
      ) {
        tasks.push(
          "NODE.md has placeholder owners — set owners to your GitHub username(s)",
        );
      }
    }
    if (repo.hasPlaceholderNode()) {
      tasks.push(
        "NODE.md has placeholder content — fill in your project's domains and description",
      );
    }
  }
  return { group: "Root Node", order: 2, tasks };
}
